applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp

```diff
@@ -0,0 +1,588 @@
+
+    auxLocalCMatrix = prod(localCMatrix, rotationMatrix);
+    rVariables.CAbsMatrix = prod(trans(rotationMatrix), auxLocalCMatrix);
+
+    for (unsigned int idim = 0; idim < TDim; ++idim) {
+            rVariables.CAbsMatrix(idim, idim) = abs(rVariables.CAbsMatrix(idim, idim));
+    }
+}
+
+/// <summary>
+/// Calculates the stiffness constants in all directions
+/// </summary>
+/// <param name="rVariables"></param>
+/// <param name="CurrentProcessInfo"></param>
+/// <param name="Geom"></param>
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::
+CalculateNodalStiffnessMatrix(NormalLysmerAbsorbingVariables& rVariables, const ProcessInfo& CurrentProcessInfo, Element::GeometryType& Geom)
+{
+    array_1d<double, 2> rStiffnessConstants;
+
+    // calculate rotation matrix
+    BoundedMatrix<double, TDim, TDim> rotationMatrix;
+    CalculateRotationMatrix(rotationMatrix, Geom);
+
+    const int local_perpendicular_direction = TDim - 1;
+
+
+    // calculate constant traction vector part
+    rStiffnessConstants[0] = rVariables.G / rVariables.virtual_thickness;
+    rStiffnessConstants[1] = rVariables.Ec / rVariables.virtual_thickness;
+
+
+    BoundedMatrix<double, TDim, TDim>             localKMatrix = ZeroMatrix(TDim, TDim);
+    BoundedMatrix<double, TDim, TDim>             auxLocalKMatrix = ZeroMatrix(TDim, TDim);
+
+    rVariables.KAbsMatrix = ZeroMatrix(TDim, TDim);
+
+    for (unsigned int idim = 0; idim < TDim; ++idim) {
+        localKMatrix(idim, idim) = rStiffnessConstants[0];
+    }
+    localKMatrix(local_perpendicular_direction, local_perpendicular_direction) = rStiffnessConstants[1];
+
+    auxLocalKMatrix = prod(localKMatrix, rotationMatrix);
+    rVariables.KAbsMatrix = prod(trans(rotationMatrix), auxLocalKMatrix);
+
+    for (unsigned int idim = 0; idim < TDim; ++idim) {
+        rVariables.KAbsMatrix(idim, idim) = abs(rVariables.KAbsMatrix(idim, idim));
+    }
+}
+
+/// <summary>
+/// This method gets the average of the variables of all the neighbour elements of the condition. 
+/// </summary>
+/// <param name="rVariables"></param>
+/// <param name="rCurrentProcessInfo"></param>
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::
+GetNeighbourElementVariables(
+    NormalLysmerAbsorbingVariables& rVariables, const ProcessInfo& rCurrentProcessInfo)
+{
+    
+    double rho_s = 0;
+    double rho_w = 0;
+    double E = 0;
+    double n = 0;
+    double nu = 0;
+    double rMeanDegreeOfSaturation = 0;
+
+    // get mean degree of saturation of all integration points in all neighbour elements
+    std::vector<double> SaturationVector;
+    
+    // get neighbour elements
+    auto neighbours = this->GetValue(NEIGHBOUR_ELEMENTS);
+    
+    int nElements = neighbours.size();
+    
+    // loop over all neighbour elements
+    int nValues = 0;
+    for (unsigned int i = 0; i < neighbours.size(); ++i) {
```

Comment:
I guess, here, averaged values in an element are calculated. These values are later used in the boundary condition (I guess). A better approach might be to map data from GPs of the element to the nodes and then interpolate it to the GPs of the boundary conditions.

Reply:
I see that the VectorOfNeighbours always has a size of 1, why is this a vector rather than a value?

Reply:
Because it is stored by "NEIGHBOUR_ELEMENTS" and this variable is defined as a vector.

Reply:
a sorry, i was confused, i thought that NEIGHBOUR_ELEMENTS was a GeoMechanics variable, rather than a kratos core variable.